✨ feat(lecture): bookmark mock data api 추가

209512 opened the pull request:
## ✅ PR 요약
- 관련 api 문서: LECT 005, 006 
- 작업 요약: 강의 북마크 기능 mock api 추가.(serializers, views, urls)

## 📄 상세 내용
- [ ] LectureBookmarkListSerializer, LectureBookmarkCreateSerializer:
    - IntegrityError 예외 처리로 동시성 문제, 중복 생성 방지
    - 북마크 생성 시 중복 체크를 위한 유효성 검사 로직(validate 메서드) 추가
- [ ]  LectureBookmarkListCreateIView: mock 데이터. 북마크 목록 조회, 추가 api. 
- [ ]  LectureBookmarkDeleteView: mock 데이터. 북마크 추가/삭제 버튼 api



## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

209512 committed:
✨ feat(lecture): bookmark serializers 추가

- 북마크 추가/삭제, 목록 조회를 위한 시리얼라이저

2joonkim commented on apps/lecture/serializers/bookmark_serializers.py:
```diff
@@ -0,0 +1,39 @@
+        # obj.lecture는 CrawledLecture 객체.
+        lecture: CrawledLecture = obj.lecture
+        total_min = lecture.duration
+
```
이 부분 라원님과 소통이 필요해 보입니다!! 꼭 참고하시고 어떻게 받는지 어떻게 크롤링하실지에 관한거 논의해보세요

Meoyoug commented on apps/lecture/serializers/bookmark_serializers.py:
```diff
@@ -0,0 +1,39 @@
+from apps.lecture.serializers.lecture_serializers import LectureListSerializer
+
+
+class LectureBookmarkButtonSerializer(serializers.ModelSerializer[LectureBookmark]):
```
이건 왜필요한걸까요?

209512 replied:
> 이건 왜필요한걸까요?

북마크 추가/삭제 api 에서 사용하려고 만들어뒀는데 혹시 필요없을까요?

Meoyoug replied:
> > 이건 왜필요한걸까요?
> 
> 북마크 추가/삭제 api 에서 사용하려고 만들어뒀는데 혹시 필요없을까요?

삭제하실 때 시리얼라이저를 사용하시나요..?

북마크 시에도 path param으로 강의 id를 넘겨받아서 북마크 객체를 생성해주면될것같은데 시리얼라이저를 굳이 거쳐야할까 생각해보시면 좋을 것 같습니다.

209512 committed:
🐛 fix(lecture): bookmark_serializers

base_crawler.py 예정에 따라 duration 함수 제거
LectureBookmarkButtonSerailizer 제거.
- 북마크 추가/삭제를 위한 별도의 Payload 없이 Path Parameter만 사용.
-> LectureBookmarkButtonSerializer는 제거하고 View에서 모델을 직접 조작.
- 추후 필요에 따라 다시 생성 후 유효성 검사 추가 가능.

209512 committed:
✨ feat(lecture): bookmark mock view, pagination mixin 추가

apps/core/utils/pagination_mixin.py:
- 공용 PaginationHandlerMixin 클래스 추가.

LectureBookmarkListAPIView에 PaginationHandlerMixin 적용.

LectureBookmarkButtonAPIView(북마크 추가/삭제): 불필요한 LectureBookmarkButtonSerializer 참조 및 사용 제거

209512 committed:
🎨 style(lecture): code formatting

Meoyoug reviewed (changes requested):
url 연결이안되어있는거 같은데 확인부탁드리고 북마크 뷰의 경우 토글뷰가 아닌 북마크 추가, 삭제를 별도 api로 구현부탁드립니다.

209512 committed:
🐛 fix(lecture): bookmark mockdata api

bookmark views
- 북마크 목록 조회, 추가를 하나의 클래스(LectureBookmarkListCreateView)로 통합
- 북마크 삭제 API는 별도의 뷰(LectureBookmarkDeleteView)로 분리.
url 수정.
Serializers.py 수정.
- IntegrityError 예외 처리로 동시성 문제, 중복 생성 방지
- 북마크 생성 시 중복 체크를 위한 유효성 검사 로직(validate 메서드) 추가

Meoyoug reviewed (changes requested):
테스트 코드 추가해서 올려주세요

209512 commented:
> 테스트 코드 추가해서 올려주세요

테스트 코드 추가하면 이 Pr이 300라인이 넘어갈 것 같은데 그래도 올려도 되나요?

Meoyoug commented:
> > 테스트 코드 추가해서 올려주세요
> 
> 테스트 코드 추가하면 이 Pr이 300라인이 넘어갈 것 같은데 그래도 올려도 되나요?

네 상관없습니다 기능개발이 됐는데 TestCode 가 없으면 다른분들이 PR 올릴때 터지잖아요

209512 committed:
✅ test(lecture): 강의 북마크 통합 테스트 코드 추가

# Conflicts:
#	apps/lecture/tests/__init__.py

209512 committed:
🎨 style(lecture): code formatting 수정

Meoyoug commented on apps/lecture/tests/test_bookmark_integration.py:
```diff
@@ -0,0 +1,179 @@
+        self.assertIn("results", response.data)
+        self.assertTrue(len(response.data["results"]) > 0)
+
+    def test_get_bookmark_list_with_page_params(self) -> None:
```
2페이지에 대한 데이터가 넘어왔는지도 확인했으면 좋을거같아요

Meoyoug commented on apps/lecture/tests/test_bookmark_integration.py:
```diff
@@ -0,0 +1,179 @@
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn("이미 북마크한 강의입니다.", response.data["detail"])
+
+    def test_post_handles_validation_error(self) -> None:
```
save 호출 시 ValidationError 가 발생하는 케이스가 있나요..?
save 호출 시점이라면 이미 Validation이 끝난 상태아닌가요?

Meoyoug commented on apps/lecture/tests/test_bookmark_integration.py:
```diff
@@ -0,0 +1,179 @@
+            self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+            self.assertIn("강제 ValidationError", str(response.data["detail"]))
+
+    def test_post_handles_generic_exception(self) -> None:
```
save 호출시 예상되는 에러로는 IntegrityError 일 것 같은데 해당 상황이 일어나도록 테스트코드를 작성해야지 모킹을 사용하는 건 적절치 않은 것 같습니다. 보통 테스트에서 모킹을 사용하는 이유는 외부 API 호출의 응답을 제어하는 용도이지 내가 작성한 코드를 모킹해서 가짜응답을 만들어내진 않아요

Meoyoug commented on apps/lecture/serializers/bookmark_serializers.py:
```diff
@@ -0,0 +1,56 @@
+        return attrs
+
+    @transaction.atomic
+    def create(self, validated_data: Dict[str, Any]) -> LectureBookmark:
```
여기서 IntegrityError는 객체 중복뿐만 아니라 schema와 다른 타입의 데이터가 들어가는 경우도 있을 수도 있구요
중복으로 단정지어서 ValidationError 를 뱉는 건 위험할 것 같습니다.

차라리 해당유저가 해당강의에 대해 북마크를 한 이력이 있는지 조회하는 것과 create 하는 것을 transaction으로 묶어서 처리했으면 이코드가 불필요했을 겁니다.

그리고 context로 create시 필요한 데이터를 전달하는 방법도 있겠지만 save호출 시 인자로 전달하는 방법도 있습니다.

원본 시리얼라이저의 save() 호출을 살펴보시면 좋겠네요.

209512 committed:
🐛 fix(lecture): 북마크 생성 시 serializer.save(user=request.user) 방식 적용

- 북마크 생성 시 user 정보를 시리얼라이저 save() 메서드 인자로 명시적으로 전달하도록 변경
- serializer에서 context["request"].user 참조 제거
- get_or_create()로 중복 생성 처리
- 뷰 POST 핸들러에서 context 전달 제거, 예외 처리 변경

209512 committed:
🐛 fix(lecture): 북마크 통합 테스트의 Mocking 제거 및 로직 일치화

- Mocking 사용 전면 제거: 테스트 코드에서 `patch`를 이용한 Mocking 사용을 모두 제거. 실제 뷰 로직 및 DB 동작에 기반한 통합 테스트로 전환. (외부 API 호출이 아닌 자가 작성 코드에 대한 Mocking 부적절 피드백 반영)

- POST IntegrityError 검증: `serializer.save()` 시 중복 북마크 생성으로 인해 발생하는 `IntegrityError`가 DRF에 의해 400 에러로 정확히 변환/처리되는지 실제 DB 조작을 통해 검증하도록 로직 수정.

- GET 페이지네이션 검증 강화: 뷰 파일의 Mock 데이터(총 15개)를 기준으로 1페이지(10개)뿐만 아니라 2페이지(5개)의 데이터 개수 및 중복 여부까지 명확히 확인하도록 테스트 보완.

- POST 에러 처리 로직 일치화: `serializer.save()` 중 발생하는 예외 처리 시, 뷰의 로직에 맞춰 테스트 기대값 수정.

Meoyoug reviewed (commented):
로컬에서 충돌 제거후에 푸시해주세요~

209512 commented:
> 로컬에서 충돌 제거후에 푸시해주세요~

넵~
